packages/ethereum/contracts/src/Announcements.sol

```diff
@@ -0,0 +1,142 @@
+    bytes32 ed25519_sig_0,
+    bytes32 ed25519_sig_1,
+    bytes32 ed25519_pub_key
+  ) external {
+    address sender_addr = address(uint160(uint256(keccak256(abi.encodePacked(secp256k1_x, secp256k1_y)))));
+
+    if (msg.sender != sender_addr) {
+      revert PublicKeyDoesNotMatchSender({pubkey: sender_addr, sender: msg.sender});
+    }
+
+    if (uint256(secp256k1_y) % 2 == 1) {
+      emit KeyBindingOdd(secp256k1_x, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key);
+    } else {
+      emit KeyBindingEven(secp256k1_x, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key);
+    }
+  }
+
+  /**
+   * [optional] Announces a IPv4 address with port for the node
+   *
+   * @dev Turns a node into a public relay node (PRN)
+   *
+   * @param ip the IPv4 address to announce
+   * @param port the port to use
+   */
+  function announce4(bytes4 ip, bytes2 port) external {
+    emit AddressAnnouncement4(msg.sender, ip, port);
+  }
+
+  /**
+   * [optional] Announces a IPv6 address with port for the node
+   *
+   * @dev Turns a node into a public relay node (PRN)
+   *
+   * @param ip the IPv6 address to announce
+   * @param port the port to use
+   */
+  function announce6(bytes16 ip, bytes2 port) external {
+    emit AddressAnnouncement6(msg.sender, ip, port);
+  }
+
+  /**
+   * Opts out from acting as a public relay node (PRN)
+   */
+  function revoke() external {
+    emit RevokeAnnouncement(msg.sender);
+  }
+
+  /**
+   * Convenience method to bind keys and announce a IPv4 address in one call.
+   */
+  function bindKeysAnnounce4(
+    bytes32 secp256k1_x,
+    bytes32 secp256k1_y,
+    bytes32 ed25519_sig_0,
+    bytes32 ed25519_sig_1,
+    bytes32 ed25519_pub_key,
+    bytes4 ip,
+    bytes2 port
+  ) external {
+    (bool successBind, ) = address(this).delegatecall(
+      abi.encodeCall(
+        HoprAnnouncements.bindKeys,
+        (secp256k1_x, secp256k1_y, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key)
+      )
+    );
+    if (successBind == false) {
+      revert();
+    }
+
+    (bool successAnnounce4, ) = address(this).delegatecall(abi.encodeCall(HoprAnnouncements.announce4, (ip, port)));
+
+    if (successAnnounce4 == false) {
+      revert();
+    }
+  }
+
+  /**
+   * Convenience method to bind keys and announce a IPv6 address in one call.
+   */
+  function bindKeysAnnounce6(
+    bytes32 secp256k1_x,
+    bytes32 secp256k1_y,
+    bytes32 ed25519_sig_0,
+    bytes32 ed25519_sig_1,
+    bytes32 ed25519_pub_key,
+    bytes16 ip,
+    bytes2 port
+  ) external {
+    (bool successBind, ) = address(this).delegatecall(
+      abi.encodeCall(
+        HoprAnnouncements.bindKeys,
+        (secp256k1_x, secp256k1_y, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key)
+      )
+    );
+    if (successBind == false) {
+      revert();
+    }
+
+    (bool successAnnounce6, ) = address(this).delegatecall(abi.encodeCall(HoprAnnouncements.announce6, (ip, port)));
+    if (successAnnounce6 == false) {
+      revert();
+    }
+  }
+}
```

Comment:
Also, as mentioned in the pm channel, it'd be ideal to provide functions that doesn't assume `msg.sender == chain_key` and verifies signature.

Reply:
added an internal function to "set" `msg.sender`